Motherships

BrettRyland opened the pull request:
The epic overhaul of AI / WM logic to allow motherships is finally ready (@SuicidalInsanity, @BillNyeTheIE):
- Overhaul of logic for which AI / WM is in control of a vessel and proper support for motherships / detachable fighters:
	- The "primary" WM is the WM closest to the root part of the vessel (an "Is Primary" label is added to the top of the WM's PAW to allow checking this).
	- Only a single AI is allowed to be active on a craft at any given time.  Activating one will deactivate the others.
	- Whenever a vessel detaches from or docks with another vessel or loses parts, the primary WM and AI automatically updates to reflect the new vessel configurations.
		- The primary AI is chosen as the first active AI sorted by hierarchical distance from the root part with the precedence: pilot, surface, VTOL, orbital.
			- E.g., if an active pilot AI is further from the root part than an active surface AI at the time that the vessels dock, the pilot AI will be chosen as the primary and the surface AI will get deactivated.
		- If no AI are active prior to docking (or on spawn), then the primary AI is chosen as the AI closest to the root part regardless of type.
		- Vessel type is also updated to reflect the AI in control of the vessel:
			- Plane
			- Ship (spaceship, boat, submarine)
			- Rover (tank/buggy, amphibious)
			- Lander (stationary)
			- Base (WM with no AI)
	- If a competition is active, then:
		- Duplicately named vessels spawned via BDA's spawning routines get a "_1"-style suffix.
		- Duplicately named fighters (vessels that detach from a parent vessel that has a WM) get a "_F1"-style suffix.
		- Vessel name deconfliction respects custom VESSELNAMING tags, only applying suffixes as necessary for deconfliction.
		- Suffixes applied to spawned vessels are persistent throughout tournaments and continuous spawn, but suffixes applied to fighters are dependent on the order that they detach from the parent.
	- Detached fighters are automatically added to active competitions / tournaments on the same team as the parent vessel and behave independently from the parent vessel.
	- If the parent craft is free or commanded to attack a location, then the fighters get the same command, otherwise they are commanded to follow the parent.
	- Ranked tournaments use the combined score of a parent vessel and its fighters when determining the ranking.
	- Access patterns for the AI / WM are optimised to be even faster than the previously used VesselModuleRegistry.

Example test craft: [Test Mothership.zip](https://github.com/user-attachments/files/21057911/Test.Mothership.zip)

Also, fixes the PID auto-tuner's "Initial Learning Rate" slider in the AI GUI.

BrettRyland committed:
WIP initial re-implementation of the motherships branch

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Update TODO list

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Add an ActiveController vessel module to non-space objects that automatically updates the WMs and AIs in order of priority when vessels are created, joined or have part count changes.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Fix AI activation/deactivation so that only the primary AI is active.

BrettRyland committed:
Rename some functions.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Sync and note

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Expand valid vessel types for competitions and reassign vessel types after updating the AI module.

BrettRyland committed:
Missed one.

BrettRyland committed:
Add an option to override which AI is primary and switch AI activation to go through that.

BrettRyland committed:
Replace calls to VesselModuleRegistry.Get...AI/WM with calls to ActiveController().WM/AI. Add FIXMEAI tags for further points that need looking into.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Replace VesselModuleRegistry.GetMissileFire calls with ActiveController calls. Flag more locally cached WM variables with FIXMEAI to investigate.

BrettRyland committed:
Make sure the targeting camera texture gets released on destroy.

BrettRyland committed:
Remove commented out calls that went via gameObject.

BrettRyland committed:
Clean-up. Add comments. Check FIXMEAI tags.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Conversion and optimisation of various AI and WM accesses.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Fix relative vs absolute paths in cts spawning log parser. If the 'logs' argument passed to the parser is a single folder, treat that as the Logs folder instead of as a file and parse the latest cts spawn log in that folder..

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Fix the PID auto-tuner's 'Initial Learning Rate' slider in the AI GUI.

BrettRyland committed:
Adapt the AI GUI to work with multiple AIs.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Fix parsing of Rect to use float values.

BrettRyland committed:
Clean up some uses of the various AI modules that weren't checking for them to be active first. Fix some performance bugs.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
When detaching a vessel, copying the parent vessel's WM and AI state.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Switch some GetModules<MissileFire> for GetMissileFires.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Tweak some debug strings.

BrettRyland committed:
WIP vessel name deconfliction.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Serialise/deserialise deconfliction data to the tournament state file to be consistent across restarts. Adjust where vessel naming gets deconflicted and where deconfliction gets reset.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Store/restore deconfliction data to the tournament state when running tournament heats.

BrettRyland committed:
Only apply vessel naming deconfliction once per vessel.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Remove the 'Lives' column in the score window when using unlimited lives in cts spawn.

BrettRyland committed:
Put logging of TeamIcons save/load behind DEBUG_OTHER checks.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge dev

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
notes

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Add IsInf and IsInfOrNaN extensions for Vector3 and Quaternion.

BrettRyland committed:
Use OnGUIWM for OnGUI stuff involving the WM to avoid repeated calculations.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
WIP vessel naming

BrettRyland committed:
Tweak camera switch changes.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Remove trailing empty space

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
If a craft gets eaten by the Kraken, check for other debris that's left the SoI and feed it to the Kraken too.

BrettRyland committed:
Fix vessel-name deconfliction for continuous spawn and continuous single spawn. Change continuous spawn to count the number of active teams instead of active craft for deciding when to spawn new craft. Account for fighters when deciding when to queue craft in continuous spawn.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Update TargetInfo class to handle changing WMs.

BrettRyland committed:
Clean up some debugging and comments. Add a null check to avoid a rare NRE.

BrettRyland committed:
Look for and remove detached fighters from spawns where the parent craft has "lost parts" as they otherwise break KSP.

BrettRyland committed:
Clean up debris that's left the SoI to avoid summoning the Kraken. If the active vessel is leaving the SoI, force-switch away from it to the first vessel in the SoI.

BrettRyland committed:
Fix tournament logic and save/load support for fighters.

BrettRyland committed:
Clean up debugging and comments.

BrettRyland committed:
Fix an issue with manually encoded quote marks in the tournament log parser not properly awarding wins.

BrettRyland committed:
Put another debug message behind the 'Debug Other' toggle.

BrettRyland committed:
Standardise (almost) all Path.Combine operations with Path.GetFullPath to remove ambiguity when comparing paths.

BrettRyland committed:
Adjust comment and tweak cts spawn logs.

BrettRyland committed:
Adjust setting the AI/WM state after detachment to be dependent on what the parent was doing.

BrettRyland committed:
Update changelog.

BrettRyland committed:
Prevent decouplers from randomly decoupling just after spawn.

BrettRyland committed:
Revert decoupler special handling and add note about some pre-active smart parts (particularly altitude and speedometer) triggering during spawning and breaking KSP.

BrettRyland committed:
Put some debugging behind a DEBUG_AI check.

BrettRyland committed:
Revert "Put some debugging behind a DEBUG_AI check." - I changed my
mind...

This reverts commit 9d35017df73ec34cffaa06d9d5ab38720267e60c.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Update TODO

BrettRyland committed:
Sync localisation

BrettRyland committed:
Fix typo in comment

SuicidalInsanity commented on BDArmory/Control/BDGenericAIBase.cs:
```diff
@@ -607,18 +597,27 @@ protected virtual void UpdateWaypoint()
+                            break;
+                        case AIType.VTOLAI:
+                            // var vtolAI = ai as BDModuleVTOLAI;
+                            // vtolAI.MaxSpeed = mSpeed > 0 ? mSpeed : originalMaxSpeed; //VTOL AI really needs expanding to actually include behavior beyond stationary hovering; could do WPs, but only if it has a fixed horizontal thrusters under IndependantThrottle...
```
VTOL AI these days does have Waypoint following support, so this could be uncommented

SuicidalInsanity commented on BDArmory/UI/BDArmoryAIGUI.cs:
```diff
@@ -296,42 +265,51 @@ void GetAIEditor()
+                currentPart = currentPart.parent;
+                ++proximity;
+            }
+            if (currentPart == null)
```
Would it be sensible to have a provision for stopping proximity checks if AI is behind a docking port or decoupler (similar to how MMGs determine what parts are the MMG and not parent craft)?

BrettRyland replied:
That's a possible option, but it means that if the AI / WM on the mothership gets destroyed, then the AI on the parasite fighters can't take over.
Also, something silly like a chain of "cars" ([Test - multicraft.zip](https://github.com/user-attachments/files/21330909/Test.-.multicraft.zip)) with different types of AIs on wouldn't be able to switch between them without undocking first, which would limit players options a little.

SuicidalInsanity commented on BDArmory/Weapons/ModuleWeapon.cs:
```diff
@@ -263,17 +263,16 @@ enum TargetAcquisitionType { None, Visual, Slaved, Radar, AutoProxy, GPS };
+                if (_weaponManager == null || !_weaponManager.IsPrimaryWM || _weaponManager.vessel != vessel)
+                    _weaponManager = vessel && vessel.loaded ? vessel.ActiveController().WM : null;
+                return _weaponManager;
             }
```
This opens up a question that I also encountered back when I was experimenting around with drones - do we want parasite craft weaponry disabled or active while still docked to the mothership? - e.g. you have a Carrier B-52-type thing, armed with a turret or 2, with a pair of parasitefighters on wing pylons each with a fixed vulcan - should the mothership be able to/have access to select the vulcan and behave as if it's fitted with the worlds most over-engineered gunpods?

BrettRyland replied:
Hmm, that's a good question that I haven't really thought about. I can see arguments (or at least, players' expectations) for having it both ways.
If they were to be restricted to only be selectable when the parasite fighter's WM is in control, then this would need to have some way of knowing when the weapon is on a parasite fighter that's detached or not, which I don't see an easy way of determining.

SuicidalInsanity replied:
Possibility A would be UpdateList() in MissileFire does a check to see if a weapon has a decoupler/docking port as a parent above it in the part tree, similar to MMGs - this would be automatic, but does have potential edge cases of weapons mounted on wings that are mounted on decouplers as a damage control for balancing lift in case of wing loss, or you have missiles mounted on Small Hardpoint/Structural Pylon parts. Treating parasite craft like MMGs would open up the possibility of automated deployment of parasite craft by the WM during combat, though.

Possibility B would be adding a toggle to ModuleWeapon/Missilebase to set if a weapon is solely a child craft weapon or accessible by all craft WMs, which could get filtered by UpdateList(), though that requires players paying attention when building. Would need to default to all weapons accessible by all WMs.

Possibility C - have a GUI along the lines of the Ammobelt tool pop up and players assign what weapons go to what WM, maybe automatically if the Editor detects multiple WMs, and saving permitted weapons to a persistent string in the WM, then UpdateList iterates through those weapons only. Again requires players configuring things, would assume weapons accessible by all WMs as the default.

SuicidalInsanity commented on BDArmory/Weapons/Missiles/MissileLauncher.cs:
```diff
@@ -1349,7 +1334,7 @@ public override void FireMissile()
             }
-            var wpm = VesselModuleRegistry.GetMissileFire(SourceVessel, true);
+            var wpm = SourceVessel.ActiveController().WM;
             if (wpm != null) Team = wpm.Team;
```
The WMs in FireMissile() would want to be the WM at time of firing, not whatever the current active WM is, to ensure proper targeting, so would warrant passing along the WM from missileFire via FireMissile(missileFire WM) instead of assigning it here. Admittedly something of an edge case as it would require a WM change at the exact same time as a launch

BrettRyland replied:
Can that change happen? All the `MissileLauncher.FireMissile()` calls are made immediately after `SendTargetDataToMissile` calls without any `yield`s, so there shouldn't be any event that switches the active WM while firing the missile.

SuicidalInsanity commented on BDArmory/Weapons/Missiles/MissileLauncher.cs:
```diff
@@ -1488,7 +1473,7 @@ IEnumerator FireReloadableMissile()
             ml.launched = true;
-            var wpm = VesselModuleRegistry.GetMissileFire(SourceVessel, true);
+            var wpm = SourceVessel.ActiveController().WM;
             ml.SourceVessel = SourceVessel;
```
Same for ReloadableFire, though this also gets into the same question as in ModuleWeapon of do we want a mother craft selecting/launching child craft missiles while child vessels are still docked?

BrettRyland replied:
Hmm, that has some `yield`s while it's spawning the missile. We could grab the WM at the start of `FireReloadableMissile` and abort if it's changed at this point, perhaps?

SuicidalInsanity replied:
That could work.

SuicidalInsanity commented on BDArmory/Weapons/Missiles/MissileLauncher.cs:
```diff
@@ -2266,9 +2251,9 @@ private void UpdateTerminalGuidance()
                         else
-                            heatTarget = BDATargetManager.GetHeatTarget(SourceVessel, vessel, new Ray(vessel.CoM, tempTargetPos - vessel.CoM), TargetSignatureData.noTarget, lockedSensorFOV / 2, heatThreshold, frontAspectHeatModifier, uncagedLock, targetCoM, lockedSensorFOVBias, lockedSensorVelocityBias, lockedSensorVelocityMagnitudeBias, lockedSensorMinAngularVelocity, SourceVessel ? VesselModule ...
+                            heatTarget = BDATargetManager.GetHeatTarget(SourceVessel, vessel, new Ray(vessel.CoM, tempTargetPos - vessel.CoM), TargetSignatureData.noTarget, lockedSensorFOV / 2, heatThreshold, frontAspectHeatModifier, uncagedLock, targetCoM, lockedSensorFOVBias, lockedSensorVelocityBias, lockedSensorVelocityMagnitudeBias, lockedSensorMinAngularVelocity, SourceVessel ? SourceVessel ...
                         if (heatTarget.exists && CheckTargetEngagementEnvelope(heatTarget.targetInfo))
```
If yes, then terminal guidance will need original time-of-launch WM for terminal guidance.

SuicidalInsanity replied:
It'd make more sense to have the ``FiredWM`` be in ``missileBase``, but that should resolve the various references to WM that should remain static in the various guidance modes that update, yes.

SuicidalInsanity commented on BDArmory/Weapons/Missiles/MissileBase.cs:
```diff
@@ -838,9 +839,9 @@ protected void UpdateHeatTarget()
                     heatTarget = BDATargetManager.GetAcousticTarget(SourceVessel, vessel, lookRay, predictedHeatTarget, lockedSensorFOV / 2, heatThreshold, targetCoM, lockedSensorFOVBias, lockedSensorVelocityBias, lockedSensorVelocityMagnitudeBias, lockedSensorMinAngularVelocity,
-                        (SourceVessel == null ? null : SourceVessel.gameObject == null ? null : SourceVessel.gameObject.GetComponent<MissileFire>()), targetVessel, IFF: hasIFF);
+                        SourceVessel ? SourceVessel.ActiveController().WM : null, targetVessel, IFF: hasIFF);
                 else
```
This should be time-of-launch WM

SuicidalInsanity commented on BDArmory/Weapons/Missiles/MissileBase.cs:
```diff
@@ -1405,7 +1406,7 @@ public Vector3d UpdateInertialTarget()
                 {
-                    var weaponManager = VesselModuleRegistry.GetMissileFire(SourceVessel);
+                    var weaponManager = SourceVessel.ActiveController().WM;
                     TargetSignatureData INStarget = TargetSignatureData.noTarget;
```
As should this.

SuicidalInsanity commented on BDArmory/Weapons/Missiles/BDModularGuidance.cs:
```diff
@@ -1599,7 +1605,7 @@ public override void FireMissile()
 
-                var wpm = VesselModuleRegistry.GetMissileFire(SourceVessel, true);
+                var wpm = SourceVessel.ActiveController().WM;
                 if (wpm != null)
```
This should be the ``weaponManager`` from ln 1595.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Fix merge.

BrettRyland committed:
Uncomment speed control for VTOL AI in BDGenericAIBase.UpdateWaypoint().

BrettRyland commented on BDArmory/UI/BDArmoryAIGUI.cs:
```diff
@@ -296,42 +265,51 @@ void GetAIEditor()
-                    }
+                    var rootPart = ship.Parts.First(); while (rootPart.parent != null) rootPart = rootPart.parent;
+                    AIs.Sort((m1, m2) => ProximityToRoot(m1.part, rootPart).CompareTo(ProximityToRoot(m2.part, rootPart))); // Same sorting as ActiveController with no active AIs.
+                    if (ActiveAI == AIs.First() as BDGenericAIBase) yield break; // It's the same AI, do nothing.
```
Good point about AIs with equal proximity to the root part. The default sorting method of `List<T>` is unstable. I'll switch them to use `Enumerable.OrderBy`, which is stable, in which case, ties will be ordered by their order in the craft file.

If the parasite craft's AI is closer to the root than the primary AI, then I think that's a PEBCAK error.

BrettRyland commented on BDArmory/Utils/VesselModuleRegistry.cs:
```diff
@@ -783,4 +859,364 @@ public void DumpRegistriesFor(Vessel vessel)
+            VesselModuleRegistry.OnVesselModified(Vessel);
+
+            // Set only the closest WM to the root part as the active WM.
+            WM = VesselModuleRegistry.GetMissileFire(Vessel);
```
It allows the WM to not be on the root part, which is often a necessity for parasite fighters, which often have a decoupler as the root part.

BrettRyland committed:
Use the right WM for setting team and updating missiles away when firing MMGs.

BrettRyland committed:
Switch sorting of modules (AI/WM) to use a stable sorting algorithm.

BrettRyland committed:
Store the WM that fired a missile for later reference and use it when updating guidance/targets. Grab the WM before decoupling in Jettison(). Questions for SI marked 'FIXME SI'.

BrettRyland committed:
Rename a variable in MultiMissileLauncher.cs and set the stored FiredByWM to that of the parent missile.

BrettRyland committed:
Add note about stable sorting for WM/AI and how ties are handled.

SuicidalInsanity commented on BDArmory/Weapons/Missiles/MissileLauncher.cs:
```diff
@@ -1591,19 +1598,19 @@ IEnumerator FireReloadableMissile()
             {
-                MissileFire datalinkwpm = ml.radarTarget.lockedByRadar.vessel.ActiveController().WM;
+                MissileFire datalinkwpm = ml.radarTarget.lockedByRadar.vessel.ActiveController().WM; // FIXME SI Does this matter if the primary WM on the datalink changes?
                 if (datalinkwpm)
```
If keeping the 'WM doesn't match, delete missile' stuff on 1475, no, because the code wouldn't execute to this point.
Otherwise, in the case that the parent ship has a radar and the child ship doesn't, it wouldn't be any different from launching a missile using a second ship's radar via a Radar Data Receiver part, and the missile's datalink WM would be set to the parent's WM for linking to the locking radar set, so this should be correct as is.

BrettRyland commented on BDArmory/Weapons/Missiles/BDModularGuidance.cs:
```diff
@@ -1264,7 +1264,7 @@ private void ResetMissile()
             MissileState = MissileStates.Idle;
             var weaponManager = WeaponManager;
-            if (weaponManager != null && weaponManager.guardFiringMissile)
+            if (weaponManager != null && weaponManager.guardFiringMissile) // FIXME SI Should this be FiredByWM?
```
It's for setups like Kurgan used in the "missile" round in RWP, where the entire craft was a dual-purpose missile/craft. It'd switch to the MMG to guide it when it was locked on, but revert to a pilot AI and reset if it missed.

BrettRyland commented on BDArmory/Weapons/Missiles/BDModularGuidance.cs:
```diff
@@ -1592,60 +1592,59 @@ public void GuiFire()
-                SetTargeting();
-                Jettison();
+            FiredByWM = WeaponManager; // Generally, the parent plane's WM, but may also be the WM on the modular missile if the missile has reset. FIXME SI If the missile has reset, should FiredByWM also get updated? What happens if a modular missile has it's own WM? Should that automatically take over when a missile is fired anyway or only on a reset? The latter seems more reasonable.
+            if (FiredByWM != null && targetVessel == null)
```
Yeah, I'm not sure if it wouldn't be better to just drop the AI/WM on MMG option.

SuicidalInsanity replied:
I can see arguments for both keeping it and removing it. 
Keeping it, you could use MMGs as a way to deploy child craft from a mothership (set it to be a GPS guidance, and have a 1 sec SmartPart timer to trigger the Reset Missile AG, AI would auto-launch fighters under full AI control as soon as a bogey enters the launch range), but that's a hacky workaround (and if going that route would be better to just commit and add full support for that without needing a Smart Part).
Removing it would clean up the code and simplify use cases (and, by extension, code maintenance/debugging), since the original reason was a hack to work around legacy AI ramming (in)accuracy and ?no longer? needed, and automating launch of fighters could be done just as easily by, say, linking fighter launch to a missile/weapon select AG.

BrettRyland replied:
I've commented out the ability to reset MMGs that have an AI/WM onboard for now.
When the MMG gets jettisoned, it'll disable any active AI/WM so that they won't interfere with anything.
We can re-enable it later if it's something people want.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland commented on BDArmory/Weapons/Missiles/MissileLauncher.cs:
```diff
@@ -1471,12 +1472,18 @@ IEnumerator FireReloadableMissile()
+                if (BDArmorySettings.DEBUG_MISSILES) Debug.Log($"[BDArmory.MissileLauncher]: Primary WM changed while spawning reloadable missile, abort firing it!");
+                if (ml is not null) Destroy(ml.gameObject); // De-spawn the missile. FIXME SI Is this sufficient?
+                yield break;
+            }
```
So, you mean to not abort the launch if the WM changes, just cache the `SourceVessel` and set it as the original one, like so?
```diff
diff --git a/BDArmory/Weapons/Missiles/MissileLauncher.cs b/BDArmory/Weapons/Missiles/MissileLauncher.cs
index a7fdc2f51..cccedce87 100644
--- a/BDArmory/Weapons/Missiles/MissileLauncher.cs
+++ b/BDArmory/Weapons/Missiles/MissileLauncher.cs
@@ -1457,2 +1457,3 @@ IEnumerator FireReloadableMissile()
             var firedByWM = SourceVessel.ActiveController().WM;
+            var sourceVessel = SourceVessel;
             part.partTransform.localScale = Vector3.zero;
@@ -1474,8 +1475,2 @@ IEnumerator FireReloadableMissile()
             }
-            if (firedByWM != SourceVessel.ActiveController().WM)
-            {
-                if (BDArmorySettings.DEBUG_MISSILES) Debug.Log($"[BDArmory.MissileLauncher]: Primary WM changed while spawning reloadable missile, abort firing it!");
-                if (ml is not null) Destroy(ml.gameObject); // De-spawn the missile. FIXME SI Is this sufficient?
-                yield break;
-            }
 
@@ -1483,3 +1478,3 @@ IEnumerator FireReloadableMissile()
             ml.launched = true;
-            ml.SourceVessel = SourceVessel;
+            ml.SourceVessel = sourceVessel;
             ml.GuidanceMode = GuidanceMode;
```

SuicidalInsanity replied:
Yes.

BrettRyland committed:
Remove resolved FIXME

BrettRyland committed:
Switch a WM for FiredByWM

BrettRyland committed:
If the WM changes while launching a missile, keep the source vessel as the original launch vessel.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Fix merge.

BrettRyland committed:
Merge branch 'dev' into multi-craft

BillNyeTheIE committed:
Added some clarification comments

- Add some clarification comments to maxOffboresight setting functions as it might be confusing for a future maintainer

BillNyeTheIE committed:
Account for Multi-Craft VRD Swapping

- Ensure multi-craft vessels add the new updated VRD
- Slight efficiency improvements, particularly ensuring `LinkAllRadars()` is called only once

BillNyeTheIE committed:
FX Improvements

- Attach bullet hit FX to the hit collider's transform (if it exists) to account for turrets and other moving parts
- Add ocean checks where they were previously missing so that bullets on planets with no oceans aren't going underwater (also for efficiency, those altitude checks aren't free!)
- Add splash effect to underwater explosions, might need tuning though!

BillNyeTheIE committed:
Account for Velocity in BulletHitFx

- Account for vessel velocity in BulletHitFX, specifically for decals applied by `ExplosionsFx` as it appears these are 1 frame off. Hit decals from projectile don't appear to require this from preliminary testing.

BillNyeTheIE committed:
Account for Missiles in Multi-Craft

- When multi-craft child vessels are launched, check if any of the parent craft's missiles are using a radar onboard the child vessel. If so, swap guidance and control of the missile to the child vessel.

BillNyeTheIE committed:
Update ChangeLog.txt

BillNyeTheIE committed:
Missile Changes + Fixes

- Remove the fixed 1 s guidance delay from ballistic and cruise guidances, these are largely unnecessary with the new controller and `maxAeroTorque` mechanics, and can be implemented via the `guidanceDelay` field. They instead prevent thrust vectoring missiles from using their thrust vectoring capabilities.
- Add `displayOrdinanceHasColliders` to disable ordinance colliders in cases like VLS cells

BillNyeTheIE committed:
Fuel Leak Vel Correction

- Theoretically, fuel leaks should have the same issues as bullet hits do.

BillNyeTheIE committed:
Explosion BulletHitFX + FuelLeakFX Correction

- Use TimeIndex instead of `fixedDeltaTime`

BillNyeTheIE committed:
Removed Nullcheck for hit.collider.transform

- hit.collider.transform should exist if hitPart exist

BillNyeTheIE committed:
Account for parts with no rb

- Perform nullcheck for existence of rb and if no rb exists, use the vessel's velocity

BillNyeTheIE committed:
Correct Mistaken Use of `MF.team` instead of `MF.teamString`

BillNyeTheIE committed:
More `MF.team` Corrections

Courtesy of DocNappers and SuicidalInsanity

BillNyeTheIE committed:
Continuous Rod Warhead Improvements

- Change AoE from 60-90° to 30-60°
- Determine if missile is above/below the target, and offset aim in that direction

BillNyeTheIE committed:
Corrected Missing Team for Explosion Bullet Hits

BillNyeTheIE committed:
Fix Dumb Mistake

- Fix missing `Mathf.Sign`

BillNyeTheIE committed:
Small Efficiency Tweak

- Use missile's local `vessel.up` instead of using `GetUpDirection` at the target position, this adjustment is only relevant once we get close to the target, thus using the missile's `vessel.up` should be fine.

BrettRyland committed:
Merge branch 'dev' into HEAD

BillNyeTheIE committed:
Update ChangeLog.txt

BrettRyland committed:
Update paintball decals to be similarly flush as bullet hit decals. Switch the stingray torpedo to use seekerTimeout instead of the deprecated radarTimeout.

BrettRyland committed:
Fix typo in comment

BrettRyland committed:
Fix decal/fuel leak/fires spawning stemming from explosions to appear in the correct place.

BrettRyland committed:
Merge commit 'refs/pull/745/head' of github.com:BrettRyland/BDArmory into multi-craft

BrettRyland committed:
Merge branch 'dev' into multi-craft

BrettRyland committed:
Disable the ability to reset MMGs with AI/WMs on them. When jettisoning a MMG, disable any AI/WM on them.

BillNyeTheIE committed:
Multi-Craft weaponChannel Mechanic

- Adds new "weaponChannel" field which determines if the current weapon manager can control said weapon

BillNyeTheIE committed:
Use Arrays for Point Defense Logic

- Replace `using VesselModuleRegistry.GetModule` calls in `PointDefenseTurretFiring` with arrays updated in `UpdateList()` which should be slightly more efficient

BillNyeTheIE committed:
Small Optimizations

- Replace several relevant `transform.position` and `targetVessel.transform.position` calls in `MissileFire` with `vessel.CoM`, replaced relevant instances of `GetUpDirection` with `vessel.up` (technically a few of the bomb ones can be replaced too, but maybe someone wants to fight on the surface of a black hole or something)

BillNyeTheIE committed:
Fix Small Issues

- Correct `pointDefenseWeaponArray` not including `dualModeAPS` weapons
- Reposition gun and weapon channel check to skip missile code for efficiency
- Fix missing `IsPrimaryWM` check in `PointDefenseTurretFiring()`

BillNyeTheIE committed:
Correct Incorrect Type

- Used `int` instead of `float`, whoops!

BillNyeTheIE committed:
Update IBDWeapon.cs

BillNyeTheIE committed:
Graceful MML Path Error Exception

- Avoid bricking MML if model path is not found

BillNyeTheIE committed:
Guard Mode Radar Range Update

- Set radar range to maximum when toggling guard mode on

BillNyeTheIE committed:
Fix MMLs Not Setting `FiredByWM`

- Fixed missing `ml.FiredByWM = FiredByWM` call which resulted in `MML` missiles not having a `FiredByWM`

BillNyeTheIE committed:
Radar Fixes

- Call `EnsureRadarData` when disconnecting radars
- Make `EnsureRadarData` not add a radar if it's disabled
- Only run rest of `RemoveRadar()` code if remove was actually successful

BillNyeTheIE committed:
Ensure MMLs Models

- Ensure MMLs get the most up to date model and scaling (if they're being used as reloadable rails)
- Also ensure MML scaling is correct (again)

BrettRyland committed:
Reset the localScale of bullet hit decals and leakFX in OnDisable to prevent Unity messing with the scale on reparenting.

BrettRyland committed:
Revert setting the transform.parent to null in OnDisable as Unity doesn't like that. Instead set it to null prior to deactivating it in the pool when force-reusing it. Also, keep that it gets set to null when manually deactivating it.

BillNyeTheIE committed:
Ensure MML Scaling (Actually)

- Forgot to include these changes in last commit, woops

BillNyeTheIE committed:
Fix MML maxOffBoresight for overrideReferenceTransform

- Disable `maxOffBoresight` sanity check via `ModuleMissileRearm` if MML has `overrideReferenceTransform` flag

BillNyeTheIE committed:
Update ChangeLog.txt

BillNyeTheIE committed:
Fix Missing Set for subMunitionPath

- Forgot to set `subMunitionPath` in `ModuleMissileRearm`

BillNyeTheIE committed:
Changelog + Targeting Camera Zoom Power

BillNyeTheIE committed:
Corrected `maxOffBoresight` Max Value + Optimization

- `maxOffBoresight` is not a cone width, nor is it used as such in the majority of the codebase. It maxes out at 180 degrees, correct code inconsistent with this.
- Small efficiency tweaks in `RadarUtils`, mostly re-use an angle if it's already calculated, use pre-calculated vectors instead of recalculating them every time. Also added some discussion on the use of `Vector3.Angle` in general.

BillNyeTheIE committed:
Targeting Camera Dynamic Zoom UI

- Zoom power UI requires a dynamic width to fit the new `Zoom X` format

BillNyeTheIE committed:
IR Missile Envelope Check

- IR missiles were missing the envelope check in `UpdateHeatTarget` which might lead to unexpected behavior. Maybe we should include a boolean flag to determine whether or not a missile will perform an envelope check in general, similar to `hasIFF`

BillNyeTheIE committed:
Adjust `maxOffBoresight` of KKV and HEKV

- Adjust `maxOffBoresight` of the two missiles to reflect the nature of `maxOffBoresight`

BillNyeTheIE committed:
Fix * instead of +

- Woops, had *= instead of +=

BillNyeTheIE committed:
Update ChangeLog.txt

- Add Sidestrafe's part changes (missiles + FLIR) to the changelog
- Separate part changes from the rest of the changes
- Delete duplicate `Conformal Decals and part variants breaking craft RCS analysis.` line in General (it already exists under Sensors)

BillNyeTheIE committed:
Update ModuleTargetingCamera.cs

BillNyeTheIE committed:
Fix Misuses of armorMass

- Fix erroneous uses of `armorMass` instead of `totalArmorQty`

BillNyeTheIE committed:
Use Armor.armorVolume

Replace equation previously used with `Armor.armorVolume` directly.

BillNyeTheIE committed:
Use ActiveController in ReportVessel

- Replace `GetMissileFires` with `ActiveController.WM`
- Only do `MissileBase` check if `ActiveController.WM` is null

BillNyeTheIE committed:
Update ProjectileUtils.cs

BrettRyland committed:
Merge branch 'dev' into multi-craft

BillNyeTheIE committed:
Fix Radar Bugs + Elevation FoV

- Introduce `elevationFOV` field to fix oversight as to radar vertical scan capabilities, set to 90 so radars won't have issues as they are now
- Fix various inefficiencies (thanks DocNappers)
- Added new `VectorUtils.GetAzimuthElevation()` function which efficiently calculates the azimuth and elevation angles (in degrees) of a vector relative to axes defined by a `forward` and an `up` vector.

BillNyeTheIE committed:
Fix Incorrect State of "Advanced Ammo Config" Button

- Fix incorrect state of "Advanced Ammo Config" button in ModuleWeapon

BillNyeTheIE committed:
Update ChangeLog.txt

BillNyeTheIE committed:
Update RadarUtils.cs

BillNyeTheIE committed:
Fix AzEl for Aspected

- Misunderstood "ReferenceTransform.up" and "ReferenceTransform.forwards". "ReferenceTransform" is rotate 90 deg.
- azAngle gets Mathf.Abs() (presumably due to symmetry?)

BillNyeTheIE committed:
Update MissileGuidance.cs

BillNyeTheIE committed:
Update MissileGuidance.cs

BillNyeTheIE committed:
Update MissileGuidance.cs

BillNyeTheIE committed:
Added More Comments

- Added more comments to describe exact conditions under which "GetAzimuthElevation()" may be used.

BillNyeTheIE committed:
Clarification Comments

More clarification comments!

BillNyeTheIE committed:
Radar Fixes + Asymmetric Radar FoVs

- Cache transform.forward/up/right etc. calls in VRD and (primarily) `ModuleRadar`.
- Allow for radars to have asymmetric azimuth and elevation limits
- Account for *some* limitations of the radar GUI (though many of asymmetric radar code is NOT accounted for and needs to be fixed, as well as proper representation of the radar scan zones in 3D space!)
- LOTS of room for improvement, many of the calculations in the GUI code could really be cached, stuff like the FoV lines and whatnot.

BillNyeTheIE committed:
Update VesselRadarData.cs

BillNyeTheIE committed:
Fix Incorrect Elevation Code + Efficiency

- GetAzlEl didn't actually get the right elevation, and neither did the code that it replaced. It got the angle of the target projected onto the forward-up plane. This has been corrected, however, this requires knowledge of the vector's magnitude.
- A number of efficiency changes have been made, replacing many instances of `Vector3.Angle` and the two now deprecated `SignedAngle` functions in `VectorUtils` and `BDAMath` (NOTE: not the `Vector3.SignedAngle` function), with `VectorUtils.GetAngleOnPlane` which uses the logic that `GetAzEl` uses to perform an efficient, `sqrt`-less signed angle measurement. NOTE: this is NOT a drop-in replacement, `SignedAngle` used a `right` vector, this uses a `left` vector, if a `right` vector is used, the result must be negated to get the right sign convention!

BillNyeTheIE committed:
Battle Damage Penetration Threshold

- Add penetration threshold to battle damage mechanics
- Clamp max penetrationFactor for battle damage, overpen isn't gonna let you do much more...

BillNyeTheIE committed:
Corrections

Many many corrections...
- Corrected many of the uses of `GetAngleOnPlane`
- Corrected bank angle on helicopters
- Corrected radar behavior
	- Radars are now assumed to be roll stabilized
	- Single (fighter-type) radar display is now roll/pitch stabilized

BillNyeTheIE committed:
Update ChangeLog.txt

BillNyeTheIE committed:
Radar Azimuth/Elevation Limit UI

- Add radar azimuth/elevation limit to the part info menu

BillNyeTheIE committed:
RWR Fixes

- No more drifting RWR positions

BillNyeTheIE committed:
Update ChangeLog.txt

BrettRyland committed:
Merge branch 'multi-craft' into HEAD

BrettRyland committed:
Add an optimised VectorUtils.Angle function and a double version of Clamp.

BrettRyland committed:
Add '.' to ends of many sentences.

BrettRyland committed:
Code simplifications.

BillNyeTheIE committed:
Revert Roll Behavior

- Revert roll behavior for BDModulePilotAI

BillNyeTheIE committed:
AI Fixes

- Fix pilot AI roll
- Fix surface AI reverse behavior

BillNyeTheIE committed:
Remove `OnGUI` `UpdateReferenceTransform` Calls

BillNyeTheIE committed:
Update VesselRadarData.cs

BrettRyland committed:
Merge branch 'dev' into HEAD

BillNyeTheIE committed:
Initial Radar GUI Refactor

- Done with the radar angles and FoVs
- Fix ModuleTargetingCamera trying to remove an incorrect timing

BrettRyland committed:
Merge commit 'refs/pull/748/head' of github.com:BrettRyland/BDArmory into HEAD

BrettRyland committed:
Fold ramming rollTarget into the if-else chain of logic.

BrettRyland committed:
Switch VectorUtils.SignedAngle to use the more optimised version of Angle.

BrettRyland committed:
Changelog tweak and formatting.

BillNyeTheIE committed:
Radar GUI Refactor

The big 'un. This one pre-calculates all of the radar contact stuff.
TODO: talk about jamming, we really ought to revamp the jamming system in general

BillNyeTheIE committed:
Fix Angles

Radar angles are all weird...

BillNyeTheIE committed:
Increase activeRadarRange for AIM-120

Ensures AIM-120 ARH will find the target within its seeker range upon terminalGuidance activation

BrettRyland committed:
Merge commit 'refs/pull/748/head' of github.com:BrettRyland/BDArmory into HEAD

BillNyeTheIE committed:
Various Fixes

- Fix RWR `referenceTransform` use, ensure the rotation is only set once per `FixedUpdate`
- Fix inefficiencies in RWR code
- Fix mistaken `guiFillScanR` line in IRST part of the code, woops.
- Fix potential use of vesselRadarData before things have properly initialized.
- Moved `VesselRadarData.UpdateGUIData` to `LateUpdate`.
- Removed redundant `vesselReferenceTransform` in `VesselRadarData`

BillNyeTheIE committed:
Improve dispRange Logic

Stop `dispRange` check from causing an out-of-bounds issue and also improve the way it's determined.

BillNyeTheIE committed:
Update VesselRadarData.cs

BrettRyland committed:
Merge commit 'refs/pull/748/head' of github.com:BrettRyland/BDArmory into HEAD

BillNyeTheIE committed:
Update VesselRadarData.cs

BillNyeTheIE committed:
Reposition `drawGUI` Check

Place `drawGUI` check in `LateUpdate()` and remove it from `FixedUpdate()` and `Update()` where it isn't really used.

BillNyeTheIE committed:
VRD + WorldToRadar Fixes

- Fix efficiency issues
- Correct missing 0.5
- Set radar max range only if we're enabling radars/IRSTs

BillNyeTheIE committed:
Fixed Log Scale

- Make radar log scaling independent of window size

BillNyeTheIE committed:
Remove `referenceTransform` Parenting

- Remove referenceTransform parenting to `vessel.ReferenceTransform`
- Null `referenceTransform` in `OnDestroy`
- Add `if (!referenceTransform)` check in `UpdateReferenceTransform()`

BrettRyland committed:
Merge commit 'refs/pull/748/head' of github.com:BrettRyland/BDArmory into HEAD

BillNyeTheIE committed:
Fixes

- Fix the log scaling issues
- Remove unnecessary range check in `MissileIsThreat`

BillNyeTheIE committed:
Fix Woopsie

- Forgot to swap radial over to 128 as well

BillNyeTheIE committed:
ARH Seeker Improvements

- Fix inefficiencies with `UpdateRadarTarget()` code
- Directly set `radarLOALSearching` to true in MissileLauncher if needed
- Fix `TargetAcquired` not being set in `radarLOALSearching` branch.

BillNyeTheIE committed:
Weave Frequency Randomization

- Add weave frequency randomization

BrettRyland committed:
Merge commit 'refs/pull/748/head' of github.com:BrettRyland/BDArmory into HEAD

BrettRyland committed:
Add an option in Team Icons to not store the team colours in the settings.cfg file. Formatting.

BrettRyland committed:
Remove redundant check.

BrettRyland committed:
Merge commit 'refs/pull/748/head' of github.com:BrettRyland/BDArmory into HEAD

BillNyeTheIE committed:
Update MissileLauncher.cs

BrettRyland committed:
Replace all (relevant) instances of Vector3.Angle with the faster, more accurate VectorUtils.Angle.

BillNyeTheIE committed:
More weaponChannel Checks

- Added `weaponChannel` checks
- Removed two `weaponChannel` checks in `PointDefenseTurretFiring()` as all missiles in `pointDefenseMissileArray` should be of the correct `weaponChannel`

BillNyeTheIE committed:
Weave Guidance Sea-Skimming Fixes

- Fix issues where the missile would overfly the target due to the use of low `gVert` values when sea-skimming.

BrettRyland committed:
Merge branch 'multi-craft' into HEAD

BrettRyland committed:
Check for a null vessel when getting a BDWarheadBase's team ID.

BillNyeTheIE committed:
Various Radar Fixes

- Fix `UpdateRangeCapability()` not also updating `rangeIndex`
- Fix `CheckRadarForLock()` not accounting for WM-less vessels
- Account for vessels about to be destroyed in the various `RadarUpdateScan` functions

BillNyeTheIE committed:
More weaponChannel Checks

BrettRyland committed:
Merge commit 'refs/pull/753/head' of github.com:BrettRyland/BDArmory into HEAD

BrettRyland committed:
Formatting.

BrettRyland committed:
Fix bug in missile debug message.

BrettRyland committed:
Add weaponChannel entry to changelog
